rbf-chunker/src/main.rs

```diff
@@ -0,0 +1,227 @@
+
+    let mut discard_start = HEADER_SIZE;
+    let mut section_start = HEADER_SIZE;
+    let mut section_end = HEADER_SIZE;
+    let mut running_crc = crc16::State::<crc16::MODBUS>::new();
+
+    while section_end != rbf.len() {
+        if section_start == section_end {
+            section_end = section_end + PACKET_MIN_SIZE;
+            if section_end > rbf.len() {
+                // footer found
+                break;
+            }
+            running_crc.update(&rbf[section_start..section_end - 2]);
+        } else if section_end - section_start < PACKET_MAX_SIZE {
+            section_end = section_end + 1;
+            running_crc.update(&[rbf[section_end - 1 - 2]]);
+        } else if section_start - discard_start >= DISCARD_MAX_SIZE {
+            section_infos.push(SectionInfo::new_unkown(
+                discard_start,
+                section_start - discard_start,
+            ));
+            return Err((section_infos, "No packet found. Refuse to continue."));
+        } else {
+            running_crc = crc16::State::<crc16::MODBUS>::new();
+            section_start = section_start + 1;
+            section_end = section_start;
+            continue;
+        }
+
+        let crc = running_crc.get();
+        let crc = [(crc & 0xFF) as u8, (crc >> 8) as u8];
+        let target_crc = [rbf[section_end - 2], rbf[section_end - 1]];
+
+        if crc == target_crc && crc != [0, 0] {
+            if section_start != discard_start {
+                section_infos.push(SectionInfo::new_unkown(
+                    discard_start,
+                    section_start - discard_start,
+                ));
+            }
+            section_infos.push(SectionInfo::new_packet(
+                section_start,
+                section_end - section_start - 2,
+                crc,
+            ));
+            section_infos.push(SectionInfo::new_packet_crc(section_end - 2));
+
+            section_start = section_end;
+            discard_start = section_end;
+            running_crc = crc16::State::<crc16::MODBUS>::new();
+        }
+    }
+
+    let (offset, size) = if discard_start == rbf.len() {
+        // Last packet was actually the footer.
+        // Discard it.
+        //
+        // Wtf???
+        eprintln!("{:?}: Discarding last packet as was footer.", id);
+        let last_packet_crc = section_infos.pop().unwrap();
+        let last_packet = section_infos.pop().unwrap();
+
+        match (last_packet_crc.stype, last_packet.stype) {
+            (SectionType::PacketCRC, SectionType::Packet { .. }) => (),
+            _ => panic!(),
+        }
+
+        (last_packet.offset, last_packet.size + 2)
+    } else {
+        (discard_start, rbf.len() - discard_start)
+    };
+
+    assert_eq!(size, FOOTER_SIZE);
+    section_infos.push(SectionInfo::new_footer(offset, size));
+
+    Ok(section_infos)
+}
+
+fn dump_sections(target_dir: &PathBuf, sections: &[SectionInfo]) -> std::io::Result<()> {
```

Comment:
I think it's more idiomatic to use `&Path` here, since `target_dir.join` allocates a new `PathBuf` anyway.